### PR TITLE
fix propagating signals on docker build

### PR DESCRIPTION
**Pull request by gy95**
**What type of PR is this?**

**What this PR does / why we need it**:
When Docker recieves SIGTERM, it forwards the signal to the container and then exits immediately. It does not wait for the container to die, it just assumes it will.
this patch will fix the shell terminal appears to have stopped the container progress, but `docker ps`  still display the container is still running.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by gy95** [comment]
```
Run make
  make
  shell: /bin/bash -e {0}
  env:
    GOROOT: /opt/hostedtoolcache/go/1.16.15/x64
hack/make-rules/build_with_container.sh hack/make-rules/build.sh 
the input device is not a TTY
make: *** [all] Error 1
Makefile:59: recipe for target 'all' failed
Error: Process completed with exit code [2](https://github.com/kubeedge/kubeedge/runs/5951108808?check_suite_focus=true#step:6:2).
```

File: hack/make-rules/build_with_container.sh [hunk 1]
```diff
@@ -26,9 +26,9 @@ KUBEEDGE_BUILD_IMAGE=${KUBEEDGE_BUILD_IMAGE:-"kubeedge/build-tools"}
 DOCKER_GID="${DOCKER_GID:-$(grep '^docker:' /etc/group | cut -f3 -d:)}"
 
 set -x
-docker run --rm -u "${UID}:${DOCKER_GID}" \
+docker run --rm -it -u "${UID}:${DOCKER_GID}" \
     --init \
```
Comment (fisherxu) [inline hack/make-rules/build_with_container.sh:30]: We can have a try if remove the `--init` here

File: hack/make-rules/build_with_container.sh [hunk 2]
```diff
@@ -26,9 +26,17 @@ KUBEEDGE_BUILD_IMAGE=${KUBEEDGE_BUILD_IMAGE:-"kubeedge/build-tools"}
 DOCKER_GID="${DOCKER_GID:-$(grep '^docker:' /etc/group | cut -f3 -d:)}"
```
Comment (fisherxu) [inline hack/make-rules/build_with_container.sh:26]:
```suggestion
DOCKER_GID="${DOCKER_GID:-$(grep '^docker:' /etc/group | cut -f3 -d:)}"
CONTAINER_RUN_OPTIONS="${CONTAINER_RUN_OPTIONS:-it}"
```
It's better to define a value :)

File: hack/make-rules/build_with_container.sh [hunk 3]
```diff
@@ -26,9 +26,17 @@ KUBEEDGE_BUILD_IMAGE=${KUBEEDGE_BUILD_IMAGE:-"kubeedge/build-tools"}
...
 
 echo "start building inside container"
-docker run --rm -u "${UID}:${DOCKER_GID}" \
+set +u
+# If in Github Action workflow or Travis CI, we disable -it option, or they will report error: the input device is not a TTY
+# In Github Action, environment variable GITHUB_ACTIONS is always set to true
+# In Travis CI, environment variable TRAVIS is always set to true
+if [[ $GITHUB_ACTIONS != true && $TRAVIS != true ]]; then
+  ENABLE_INTERACTIVE_TTY="-it"
```
Comment (fisherxu) [inline hack/make-rules/build_with_container.sh:34]:
```suggestion
  CONTAINER_RUN_OPTIONS=""
```

File: hack/make-rules/build_with_container.sh [hunk 4]
```diff
@@ -26,9 +26,17 @@ KUBEEDGE_BUILD_IMAGE=${KUBEEDGE_BUILD_IMAGE:-"kubeedge/build-tools"}
 DOCKER_GID="${DOCKER_GID:-$(grep '^docker:' /etc/group | cut -f3 -d:)}"
 
 echo "start building inside container"
-docker run --rm -u "${UID}:${DOCKER_GID}" \
+set +u
+# If in Github Action workflow or Travis CI, we disable -it option, or they will report error: the input device is not a TTY
+# In Github Action, environment variable GITHUB_ACTIONS is always set to true
+# In Travis CI, environment variable TRAVIS is always set to true
+if [[ $GITHUB_ACTIONS != true && $TRAVIS != true ]]; then
```
Comment (fisherxu) [inline hack/make-rules/build_with_container.sh:33]:
```suggestion
if [[ $GITHUB_ACTIONS = true || $TRAVIS = true ]]; then
```
for ci scenario, disable it.

Comment (gy95) [reply hack/make-rules/build_with_container.sh]: fixed all above

File: hack/make-rules/build_with_container.sh [hunk 5]
```diff
@@ -24,11 +24,20 @@ KUBEEDGE_ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")/../.." && pwd -P)"
...
 
 echo "start building inside container"
-docker run --rm -u "${UID}:${DOCKER_GID}" \
+set +u
+# If in Github Action workflow or Travis CI, we disable -it option, or they will report error: the input device is not a TTY
+# In Github Action, environment variable GITHUB_ACTIONS is always set to true
+# In Travis CI, environment variable TRAVIS is always set to true
+if [[ $GITHUB_ACTIONS == true || $TRAVIS == true ]]; then
+  CONTAINER_RUN_OPTIONS=""
+fi
```
Comment (fisherxu) [inline hack/make-rules/build_with_container.sh:36]: How about set `CONTAINER_RUN_OPTIONS` to "" in github action and travis? So code can independent to the ci.

File: hack/make-rules/build_with_container.sh [hunk 6]
```diff
@@ -24,11 +24,20 @@ KUBEEDGE_ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")/../.." && pwd -P)"
 MOUNTPATH="${MOUNTPATH:-/kubeedge}"
 KUBEEDGE_BUILD_IMAGE=${KUBEEDGE_BUILD_IMAGE:-"kubeedge/build-tools"}
 DOCKER_GID="${DOCKER_GID:-$(grep '^docker:' /etc/group | cut -f3 -d:)}"
+CONTAINER_RUN_OPTIONS="${CONTAINER_RUN_OPTIONS:--it}"
 
 echo "start building inside container"
-docker run --rm -u "${UID}:${DOCKER_GID}" \
+set +u
```
Comment (fisherxu) [inline hack/make-rules/build_with_container.sh:30]: Is the `set +u` needed here?

Comment (gy95) [reply hack/make-rules/build_with_container.sh:30]: yes, or will report error: hack/make-rules/build_with_container.sh: line 34: GITHUB_ACTIONS: unbound variable

File: .travis.yml [hunk 7]
```diff
@@ -15,6 +15,7 @@ go:
 
 env:
   - HOME=/home/travis
+  - CONTAINER_RUN_OPTIONS=" "
```
Comment (fisherxu) [inline .travis.yml:18]:
```suggestion
  - HOME=/home/travis CONTAINER_RUN_OPTIONS=""
```
Keep in one line :)

File: .github/workflows/main.yaml [hunk 8]
```diff
@@ -7,6 +7,9 @@ on:
     tags:
   pull_request:
 
+env:
+  CONTAINER_RUN_OPTIONS: " "
```
Comment (fisherxu) [inline .github/workflows/main.yaml:11]:
```suggestion
  CONTAINER_RUN_OPTIONS: ""
```

Comment (gy95) [reply .github/workflows/main.yaml:11]: if fixed like this, this means env is not set,  and will be set to default -it

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**